compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -1729,51 +1730,58 @@ fn receiver_is_valid<'tcx>(
         autoderef = autoderef.include_raw_pointers();
     }
 
+    // The first type is `receiver_ty`, which we know its not equal to `self_ty`; skip it.
+    autoderef.next();
+
     let receiver_trait_def_id = tcx.require_lang_item(LangItem::Receiver, Some(span));
 
     // Keep dereferencing `receiver_ty` until we get to `self_ty`.
-    while let Some((potential_self_ty, _)) = autoderef.next() {
-        debug!(
-            "receiver_is_valid: potential self type `{:?}` to match `{:?}`",
-            potential_self_ty, self_ty
-        );
+    loop {
+        if let Some((potential_self_ty, _)) = autoderef.next() {
+            debug!(
+                "receiver_is_valid: potential self type `{:?}` to match `{:?}`",
+                potential_self_ty, self_ty
+            );
 
-        // Check if the self type unifies. If it does, then commit the result
-        // since it may have region side-effects.
-        if let Ok(()) = wfcx.infcx.commit_if_ok(|_| {
-            let ocx = ObligationCtxt::new(wfcx.infcx);
-            ocx.eq(&cause, wfcx.param_env, self_ty, potential_self_ty)?;
-            if ocx.select_all_or_error().is_empty() { Ok(()) } else { Err(NoSolution) }
-        }) {
-            wfcx.register_obligations(autoderef.into_obligations());
-            return true;
-        }
+            if can_eq_self(potential_self_ty) {
+                wfcx.register_obligations(autoderef.into_obligations());
+
+                if let Err(err) = wfcx.eq(&cause, wfcx.param_env, self_ty, potential_self_ty) {
+                    infcx
+                        .err_ctxt()
+                        .report_mismatched_types(&cause, self_ty, potential_self_ty, err)
+                        .emit();
+                }
 
-        // Without `feature(arbitrary_self_types)`, we require that each step in the
-        // deref chain implement `receiver`.
-        if !arbitrary_self_types_enabled {
-            if !receiver_is_implemented(
-                wfcx,
-                receiver_trait_def_id,
-                cause.clone(),
-                potential_self_ty,
-            ) {
-                // We cannot proceed.
                 break;
+            } else {
+                // Without `feature(arbitrary_self_types)`, we require that each step in the
+                // deref chain implement `receiver`
+                if !arbitrary_self_types_enabled
+                    && !receiver_is_implemented(
+                        wfcx,
+                        receiver_trait_def_id,
+                        cause.clone(),
+                        potential_self_ty,
+                    )
+                {
+                    return false;
+                }
             }
-
-            // Register the bound, in case it has any region side-effects.
-            wfcx.register_bound(
-                cause.clone(),
-                wfcx.param_env,
-                potential_self_ty,
-                receiver_trait_def_id,
-            );
```

Comment:
This may have been perf negative, since we now have trait goals to process... If this is the problem, then we probably need to accept this for soundness reasons if we ever want to stabilize receiver lol

Reply:
so this means we only check `T: Receiver` ignoring regions? :thinking:

Reply:
As we were previously. Let me try adding it back and seeing what happens still...